### DISCUSSION TITLE
How to override the sqlite (driver: pysqlite) datetime string serialization to the DB ?

### DISCUSSION THREAD

**Discussion by vtgn** [sqlalchemy/sqlalchemy#12138] [Usage Questions]
Hi!

I would like to store my datetime into the ISO UTC Zulu format (\<YYYY\>-\<MM\>-\<DD\>T\<HH\>:\<MM\>:\<SS\>.\<sss\>Z) in my sqlite DB.
But I can't found how to override this serialization, because even if I use a TypeDecorator, SQLAlchemy still wants a Python datetime instance returned by the implemented `process_bind_param` method, and inside the INSERT command generated by SqlAlchemyn the date is always to the format '\<YYYY\>-\<MM\>-\<DD\> \<HH\>:\<mm\>:\<SS\>.\<ssssss\>' which is not what I want.

Regards.

**Comment by zzzeek** [comment]
SQLite's datetime format is fully customizable by itself and you dont need TypeDecorator for that, see https://docs.sqlalchemy.org/en/20/dialects/sqlite.html#sqlalchemy.dialects.sqlite.DATETIME

**Reply by vtgn** [reply]
OMGGGGG the noob !!! \>_\< !!! I searched and tried everything possible for hours... except to see the data type itself... and the solution was so simple, I'm so depressed... :/

Thank you very much.

Regards.

**Reply by vtgn** [reply]
@zzzeek 
Hummm... finally this is not so easy to do... it seems the TIMESTAMP type doesn't specify  a `storage_format` parameter, and if we use it, it seems it uses the DATETIME type behaviour which allows to specify a `storage_format`.

My software must be compatible with different dialects, so I must implement a TypeDecorator to map the TIMESTAMP type to the DATETIME type for sqlite dialect only, and set my expected `storage_format`. But it seems it's impossible to give a `storage_format` with milliseconds instead of microseconds.

Here is the implementation in the `sqlite.DATETIME` class:
```python
    def bind_processor(self, dialect):
        datetime_datetime = datetime.datetime
        datetime_date = datetime.date
        format_ = self._storage_format

        def process(value):
            if value is None:
                return None
            elif isinstance(value, datetime_datetime):
                return format_ % {
                    "year": value.year,
                    "month": value.month,
                    "day": value.day,
                    "hour": value.hour,
                    "minute": value.minute,
                    "second": value.second,
                    "microsecond": value.microsecond,
                }
            elif isinstance(value, datetime_date):
                return format_ % {
                    "year": value.year,
                    "month": value.month,
                    "day": value.day,
                    "hour": 0,
                    "minute": 0,
                    "second": 0,
                    "microsecond": 0,
                }
            else:
                raise TypeError(
                    "SQLite DateTime type only accepts Python "
                    "datetime and date objects as input."
                )

        return process
```

I need to replace the `microsecond` part by the integer part of its division by 1000. Is it possible by simply set the `storage_format` parameter of DATETIME, or it needs to override the `bind_processor` method of my `TypeDecorator`?

The documentation of `bind_processor` says:

> User-defined subclasses of _types.TypeDecorator should not implement this method, and should instead implement _types.TypeDecorator.process_bind_param so that the "inner" processing provided by the implementing type is maintained.

I totally don't understand why this documentation says that, when the `process_bind_param` can't return a str type, otherwise you get this error during execution:

> TypeError: SQLite DateTime type only accepts Python datetime and date objects as input.

So it's seems the overriding of `bind_processor` is required.

Here is the current version of my TypeDecorator:

```python
class MyCustomTimestamp(TypeDecorator):
    impl = TIMESTAMP(timezone=True)

    cache_ok = True

    @override
    def load_dialect_impl(self, dialect: Dialect) -> TypeEngine[Any]:
        if dialect.name == "sqlite":
            return sqlite.DATETIME(
                timezone=True,
                storage_format=(
                    "%(year)04d-%(month)02d-%(day)02dT%(hour)02d:%(minute)02d:%(second)02d."
                    "%(microsecond)06dZ"
                ),
            )

        return super().load_dialect_impl(dialect)
```

I tried to set `%(microsecond)03d`, but it still generates 6 digits.

Thanks.

Regards.

**Reply by CaselIT** [reply]
> My software must be compatible with different dialects, so I must implement a TypeDecorator to map the TIMESTAMP type to the DATETIME type for sqlite dialect 

There's no need to use a type decorator for that. You can use with_variant: https://docs.sqlalchemy.org/en/20/core/type_api.html#sqlalchemy.types.TypeEngine.with_variant

**Reply by vtgn** [reply]
> There's no need to use a type decorator for that. You can use with_variant: https://docs.sqlalchemy.org/en/20/core/type_api.html#sqlalchemy.types.TypeEngine.with_variant

Ok, thanks for the tip, but now how to generate the format I want?

**Reply by vtgn** [reply]
Finally it worked by overriding the `bind_processor` method, even if the documentation disapproves it.
Here is my working result:

```python
class MyCustomTimestamp(TypeDecorator):
    impl = TIMESTAMP(timezone=True)

    cache_ok = True

    @override
    def bind_processor(self, dialect: Dialect) -> _BindProcessorType | None:
        if dialect.name == "sqlite":

            format: Final[str] = (
                "%(year)04d-%(month)02d-%(day)02dT%(hour)02d:%(minute)02d:%(second)02d."
                "%(millisecond)03dZ"
            )

            def process(value):
                if value is None:
                    return None
                if isinstance(value, datetime):
                    millisecond: int = round(value.microsecond / 1000)
                    if millisecond == 1000:
                        value = value + timedelta(seconds=1)
                        value = value.replace(microsecond=0)
                        millisecond = 0
                    return format_ % {
                        "year": value.year,
                        "month": value.month,
                        "day": value.day,
                        "hour": value.hour,
                        "minute": value.minute,
                        "second": value.second,
                        "millisecond": millisecond,
                        "microsecond": value.microsecond,
                    }

                raise TypeError(
                    "SQLite TIMESTAMP type only accepts Python datetime."
                )

            return process

        return super().bind_processor(dialect)
```
I don't need to change TIMESTAMP by DATETIME.
What do you think about my solution?

I think SqlAlchemy could improve this by including like I did the possibility to set the millisecond field, and to enable to give a `storage_format` to the sqlite.TIMESTAMP datatype constructor.

Regards.

**Reply by CaselIT** [reply]
> Ok, thanks for the tip, but now how to generate the format I want?

You seem to have a very peculiar need that's not available by default

>  think SqlAlchemy could improve this by including like I did the possibility to set the millisecond field, and to enable to give a storage_format to the sqlite.TIMESTAMP datatype constructor.

I don't think this is a common use case. I'm not sure why you need it

**Reply by vtgn** [reply]
The ISO UTC Zulu format with 3 digits for milliseconds is a well known standard. I decided to store this format because it is easily human readable, and avoid the problems of errors when reading and sharing between different applications because of the timezones. This is also the format I require and generate in my http interfaces.
Regards.

**Reply by vtgn** [reply]
However, subclassing `TypeDecorator` requires to override 4 methods: `process_bind_param`, `process_literal_param`, `process_result_value`, `python_type`.
If I don't implement them, it works perfectly as expected. But if I try to do it, I got an unexpected error after the execution of process_result_value.
FYI:
- in `process_bind_param` I return directly the parameter value
- in `process_literal_param` I return the conversion of the datetime typed value param to the ISO UTC Zulu string format.
- in `process_result_value` I return the conversion of the ISO UTC Zulu string format to a datetime instance.
- in `python_type` I return datetime class.

Where am I wrong? How to override them and get the same behavior as if they are not implemented?

Regards.

**Reply by CaselIT** [reply]
That format should be compatible with the default iso format, so `process_result_value` may not be needed.
If not can you provide a runnable example that shows the error you mentioned?

**Reply by zzzeek** [reply]
Here is a complete round trip with your requested format using only "storage_format" and "with_variant()", illustrated in SQLite and PostgreSQL with the same model

```py
from __future__ import annotations

import datetime

from sqlalchemy import create_engine
from sqlalchemy import select
from sqlalchemy import text
from sqlalchemy import TIMESTAMP
from sqlalchemy.dialects import sqlite
from sqlalchemy.orm import DeclarativeBase
from sqlalchemy.orm import Mapped
from sqlalchemy.orm import mapped_column
from sqlalchemy.orm import Session

zulu_datetime = sqlite.DATETIME(
    # this format when parsed by fromisoformat() will give us a tz-aware datetime,
    # so use timezone=True
    storage_format=(
        "%(year)04d-%(month)02d-%(day)02dT"
        "%(hour)02d:%(minute)02d:%(second)02d.%(microsecond)03dZ"
    ),
    timezone=True,
)


class Base(DeclarativeBase):
    pass


class A(Base):
    __tablename__ = "a"

    id: Mapped[int] = mapped_column(primary_key=True)
    data: Mapped[datetime.datetime] = mapped_column(
        TIMESTAMP(timezone=True).with_variant(zulu_datetime, "sqlite")
    )


for url in "sqlite://", "postgresql://scott:tiger@localhost/test":
    e = create_engine(url, echo=True)
    Base.metadata.drop_all(e)
    Base.metadata.create_all(e)

    s = Session(e)

    # note for full round trip we cannot store microseconds less than 1000
    # since the storage format only stores milliseconds
    the_date = datetime.datetime(2024, 12, 2, 15, 30, 55, 289000).astimezone(
        datetime.UTC
    )
    s.add(A(data=the_date))
    s.commit()

    assert s.scalar(select(A.data)) == the_date

    if e.name == "sqlite":
        # assert format
        assert s.scalar(text("select data FROM a")) == '2024-12-02T20:30:55.289000Z'
```

**Reply by zzzeek** [reply]
here is a second version that creates a simple subclass of sqlite.DATETIME in order to store exactly milliseconds in three digits

```py
from __future__ import annotations

import datetime

from sqlalchemy import create_engine
from sqlalchemy import select
from sqlalchemy import text
from sqlalchemy import TIMESTAMP
from sqlalchemy.dialects import sqlite
from sqlalchemy.orm import DeclarativeBase
from sqlalchemy.orm import Mapped
from sqlalchemy.orm import mapped_column
from sqlalchemy.orm import Session


class SLMilliDatetime(sqlite.DATETIME):

    @property
    def format_is_text_affinity(self):
        return True

    def bind_processor(self, dialect):
        datetime_datetime = datetime.datetime
        datetime_date = datetime.date
        format_ = self._storage_format

        def process(value):
            if value is None:
                return None
            elif isinstance(value, datetime_datetime):
                return format_ % {
                    "year": value.year,
                    "month": value.month,
                    "day": value.day,
                    "hour": value.hour,
                    "minute": value.minute,
                    "second": value.second,
                    "millisecond": value.microsecond / 1000,
                }
            elif isinstance(value, datetime_date):
                return format_ % {
                    "year": value.year,
                    "month": value.month,
                    "day": value.day,
                    "hour": 0,
                    "minute": 0,
                    "second": 0,
                    "millisecond": 0,
                }
            else:
                raise TypeError(
                    "SQLite DateTime type only accepts Python "
                    "datetime and date objects as input."
                )

        return process

zulu_datetime = SLMilliDatetime(
    # this format when parsed by fromisoformat() will give us a tz-aware datetime,
    # so use timezone=True
    storage_format=(
        "%(year)04d-%(month)02d-%(day)02dT%(hour)02d:%(minute)02d:%(second)02d."
                "%(millisecond)03dZ"
    ),
    timezone=True,
)


class Base(DeclarativeBase):
    pass


class A(Base):
    __tablename__ = "a"

    id: Mapped[int] = mapped_column(primary_key=True)
    data: Mapped[datetime.datetime] = mapped_column(
        TIMESTAMP(timezone=True).with_variant(zulu_datetime, "sqlite")
    )


for url in "sqlite://", "postgresql://scott:tiger@localhost/test":
    e = create_engine(url, echo=True)
    Base.metadata.drop_all(e)
    Base.metadata.create_all(e)

    s = Session(e)

    # note for full round trip we cannot store microseconds less than 1000
    # since the storage format only stores milliseconds
    the_date = datetime.datetime(2024, 12, 2, 15, 30, 55, 289000).astimezone(
        datetime.UTC
    )
    s.add(A(data=the_date))
    s.commit()

    assert s.scalar(select(A.data)) == the_date

    if e.name == "sqlite":
        # assert format
        assert s.scalar(text("select data FROM a")) == '2024-12-02T20:30:55.289Z'
```

**Reply by vtgn** [reply]
> Here is a complete round trip with your requested format using only "storage_format" and "with_variant()", illustrated in SQLite and PostgreSQL with the same model
> 
> ```python
> from __future__ import annotations
> 
> import datetime
> 
> from sqlalchemy import create_engine
> from sqlalchemy import select
> from sqlalchemy import text
> from sqlalchemy import TIMESTAMP
> from sqlalchemy.dialects import sqlite
> from sqlalchemy.orm import DeclarativeBase
> from sqlalchemy.orm import Mapped
> from sqlalchemy.orm import mapped_column
> from sqlalchemy.orm import Session
> 
> zulu_datetime = sqlite.DATETIME(
>     # this format when parsed by fromisoformat() will give us a tz-aware datetime,
>     # so use timezone=True
>     storage_format=(
>         "%(year)04d-%(month)02d-%(day)02dT"
>         "%(hour)02d:%(minute)02d:%(second)02d.%(microsecond)03dZ"
>     ),
>     timezone=True,
> )
> 
> 
> class Base(DeclarativeBase):
>     pass
> 
> 
> class A(Base):
>     __tablename__ = "a"
> 
>     id: Mapped[int] = mapped_column(primary_key=True)
>     data: Mapped[datetime.datetime] = mapped_column(
>         TIMESTAMP(timezone=True).with_variant(zulu_datetime, "sqlite")
>     )
> 
> 
> for url in "sqlite://", "postgresql://scott:tiger@localhost/test":
>     e = create_engine(url, echo=True)
>     Base.metadata.drop_all(e)
>     Base.metadata.create_all(e)
> 
>     s = Session(e)
> 
>     # note for full round trip we cannot store microseconds less than 1000
>     # since the storage format only stores milliseconds
>     the_date = datetime.datetime(2024, 12, 2, 15, 30, 55, 289000).astimezone(
>         datetime.UTC
>     )
>     s.add(A(data=the_date))
>     s.commit()
> 
>     assert s.scalar(select(A.data)) == the_date
> 
>     if e.name == "sqlite":
>         # assert format
>         assert s.scalar(text("select data FROM a")) == '2024-12-02T20:30:55.289000Z'
> ```

It's invalid: there are 6 digits for microseconds. I've already tried `%(microsecond)03d` and it returns 6 digits.

**Reply by vtgn** [reply]
> here is a second version that creates a simple subclass of sqlite.DATETIME in order to store exactly milliseconds in three digits
> 
> ```python
> from __future__ import annotations
> 
> import datetime
> 
> from sqlalchemy import create_engine
> from sqlalchemy import select
> from sqlalchemy import text
> from sqlalchemy import TIMESTAMP
> from sqlalchemy.dialects import sqlite
> from sqlalchemy.orm import DeclarativeBase
> from sqlalchemy.orm import Mapped
> from sqlalchemy.orm import mapped_column
> from sqlalchemy.orm import Session
> 
> 
> class SLMilliDatetime(sqlite.DATETIME):
> 
>     @property
>     def format_is_text_affinity(self):
>         return True
> 
>     def bind_processor(self, dialect):
>         datetime_datetime = datetime.datetime
>         datetime_date = datetime.date
>         format_ = self._storage_format
> 
>         def process(value):
>             if value is None:
>                 return None
>             elif isinstance(value, datetime_datetime):
>                 return format_ % {
>                     "year": value.year,
>                     "month": value.month,
>                     "day": value.day,
>                     "hour": value.hour,
>                     "minute": value.minute,
>                     "second": value.second,
>                     "millisecond": value.microsecond / 1000,
>                 }
>             elif isinstance(value, datetime_date):
>                 return format_ % {
>                     "year": value.year,
>                     "month": value.month,
>                     "day": value.day,
>                     "hour": 0,
>                     "minute": 0,
>                     "second": 0,
>                     "millisecond": 0,
>                 }
>             else:
>                 raise TypeError(
>                     "SQLite DateTime type only accepts Python "
>                     "datetime and date objects as input."
>                 )
> 
>         return process
> 
> zulu_datetime = SLMilliDatetime(
>     # this format when parsed by fromisoformat() will give us a tz-aware datetime,
>     # so use timezone=True
>     storage_format=(
>         "%(year)04d-%(month)02d-%(day)02dT%(hour)02d:%(minute)02d:%(second)02d."
>                 "%(millisecond)03dZ"
>     ),
>     timezone=True,
> )
> 
> 
> class Base(DeclarativeBase):
>     pass
> 
> 
> class A(Base):
>     __tablename__ = "a"
> 
>     id: Mapped[int] = mapped_column(primary_key=True)
>     data: Mapped[datetime.datetime] = mapped_column(
>         TIMESTAMP(timezone=True).with_variant(zulu_datetime, "sqlite")
>     )
> 
> 
> for url in "sqlite://", "postgresql://scott:tiger@localhost/test":
>     e = create_engine(url, echo=True)
>     Base.metadata.drop_all(e)
>     Base.metadata.create_all(e)
> 
>     s = Session(e)
> 
>     # note for full round trip we cannot store microseconds less than 1000
>     # since the storage format only stores milliseconds
>     the_date = datetime.datetime(2024, 12, 2, 15, 30, 55, 289000).astimezone(
>         datetime.UTC
>     )
>     s.add(A(data=the_date))
>     s.commit()
> 
>     assert s.scalar(select(A.data)) == the_date
> 
>     if e.name == "sqlite":
>         # assert format
>         assert s.scalar(text("select data FROM a")) == '2024-12-02T20:30:55.289Z'
> ```

That's what I did above: override `bind_processor`. But I must use `TypeDecorator` and not a sqlite only type because the client required to be able to implement easily the use of another dialect, so the types used to type the models fields must be the same and valid for any dialect. `TypeDecorator` is perfect for this, because we can override it by another for any dialect.

**Reply by vtgn** [reply]
> That format should be compatible with the default iso format, so `process_result_value` may not be needed. If not can you provide a runnable example that shows the error you mentioned?

Static type linters require to override these 4 methods that are unimplemented in the TypeDecorator class.

**Reply by CaselIT** [reply]
What type linter is this? The TypeDecorator has such method declared, so I think that count as a bug in the type linter. You can safely not implement them

**Reply by vtgn** [reply]
> What type linter is this? The TypeDecorator has such method declared, so I think that count as a bug in the type linter. You can safely not implement them

This is Pylint. I was surprised too that it can notifies such things. I supposed this is its behaviour when a `NotImplementedError` is raised.

> Method 'process_bind_param' is abstract in class 'TypeDecorator' but is not overridden in child class 'IsoUtcZuluTimestamp' Pylint(W0223:abstract-method)
Method 'process_literal_param' is abstract in class 'TypeDecorator' but is not overridden in child class 'IsoUtcZuluTimestamp' Pylint(W0223:abstract-method)
Method 'process_result_value' is abstract in class 'TypeDecorator' but is not overridden in child class 'IsoUtcZuluTimestamp' Pylint(W0223:abstract-method)
Method 'python_type' is abstract in class 'TypeEngine' but is not overridden in child class 'IsoUtcZuluTimestamp' Pylint(W0223:abstract-method)
Method 'process_bind_param' is abstract in class 'TypeDecorator' but is not overridden in child class 'IsoUtcZuluTimestamp' Pylint(W0223:abstract-method)
Method 'process_literal_param' is abstract in class 'TypeDecorator' but is not overridden in child class 'IsoUtcZuluTimestamp' Pylint(W0223:abstract-method)
Method 'process_result_value' is abstract in class 'TypeDecorator' but is not overridden in child class 'IsoUtcZuluTimestamp' Pylint(W0223:abstract-method)
Method 'python_type' is abstract in class 'TypeEngine' but is not overridden in child class 'IsoUtcZuluTimestamp' Pylint(W0223:abstract-method)

**Reply by CaselIT** [reply]
I guess sqlalchemy is not compatible with pylint. There are many place were `NotImplementedError` to mark "this method is optional", not it's abstract

**Reply by vtgn** [reply]
> I guess sqlalchemy is not compatible with pylint. There are many place were `NotImplementedError` to mark "this method is optional", not it's abstract

And Pylint doesn't generate this warning when the methods are really abstract (with no implementation).

The linters often apply strange specifications. Like MyPy which considers the `Type` typed objects as non abstract class. What a bad decision...
